ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -83,34 +98,32 @@ muxStart udesc wfn rfn sdufn close  rescb_m = do
                         | ptcl <- [minBound..maxBound]
                         , mode <- [ModeInitiator, ModeResponder] ]
 
-    mpsJob cnt pmss mpd = do
+    mpsJob cnt pmss udesc mpdId = do
+        let mpd = udesc mpdId
         w_i <- atomically newEmptyTMVar
         w_r <- atomically newEmptyTMVar
 
-        return [ mpdInitiator mpd (muxChannel pmss (mpdId mpd) ModeInitiator w_i cnt)
+        return [ mpdInitiator mpd (muxChannel pmss (AppProtocolId mpdId) ModeInitiator w_i cnt)
                      >> mpsJobExit cnt
-               , mpdResponder mpd (muxChannel pmss (mpdId mpd) ModeResponder w_r cnt)
+               , mpdResponder mpd (muxChannel pmss (AppProtocolId mpdId) ModeResponder w_r cnt)
                      >> mpsJobExit cnt
                ]
 
     -- cnt represent the number of SDUs that are queued but not yet sent.
     -- job threads will be prevented from exiting until all SDUs have been transmitted.
-    mpsJobExit cnt = atomically $ do
-        c <- readTVar cnt
-        unless (c == 0) retry
+    mpsJobExit cnt = do
+        muxBearerSetState bearer Dying
+        atomically $ do
+            c <- readTVar cnt
+            unless (c == 0) retry
 
-muxControl :: (MonadSTM m, MonadSay m, Ord ptcl, Enum ptcl) =>
-    PerMuxSharedState ptcl m ->
-    MiniProtocolMode ->
-    m ()
+muxControl :: (HasCallStack, MonadSTM m, MonadSay m, MonadThrow m, Ord ptcl, Enum ptcl)
+           => PerMuxSharedState ptcl m
+           -> MiniProtocolMode
+           -> m ()
 muxControl pmss md = do
-    w <- atomically newEmptyTMVar
-    forever $ do
-        -- XXX actual protocol is missing
-        blob <- atomically $ readTBQueue (ingressQueue (dispatchTable pmss) Muxcontrol md)
-        --say $ printf "muxcontrol mode %s blob len %d" (show md) (BL.length blob)
-        atomically $ putTMVar w blob
-        atomically $ writeTBQueue (tsrQueue pmss) (TLSRDemand Muxcontrol md (Wanton w))
+    _ <- atomically $ readTBQueue (ingressQueue (dispatchTable pmss) Muxcontrol md)
```

Comment:
This is strange, we read the queue and we dump the result.  Why do we want to block until the ingress queue is non empty? Could you explain this.

Reply:
At the moment the only valid MuxControl messages are sent during version negotiation. To send them is an error and will cause the Mux to be torn down.

This may change with the addition of a clean shutdown message in the future.

Reply:
Could we use `Muxcontrol` in `killNode` to do the clean shutdown? Maybe this way we wouldn't need to track all the `Async`s.

Reply:
Yes, for a clean shutdown there would have to be an interface between `Muxcontrol` and kill `killNode`. I think it is fine for `muxStart` to track the asyncs it starts, where it gets a bit iffy is in the `accept` case for sockets where we spawn an unknown number of MuxBearers with their corresponding threads.

Reply:
I think we will not need to track those.  I added `Connection` type and `withConnection` combinator which allows to run a connection and handle the resources well (close the socket) .  Handling this locally, means we don't need to be clever inside the library.

Reply:
https://github.com/input-output-hk/ouroboros-network/pull/359/commits/cbc3f152d78e87b573cca8a6ae4855073eb7b358#diff-7237e50e12d08b9ba55187257f498235R160